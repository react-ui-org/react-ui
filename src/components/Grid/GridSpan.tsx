import React from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { transferProps } from '../../helpers/transferProps';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { generateResponsiveCustomProperties } from './_helpers/generateResponsiveCustomProperties';
import styles from './Grid.module.scss';
import { GridSpanProps } from './GridSpan.types';

export const GridSpan: React.FunctionComponent<GridSpanProps> = ({
  children,
  columns = 1,
  rows = 1,
  tag: Tag = 'div',
  ...restProps
}) => {
  if (isChildrenEmpty(children)) {
    return null;
  }

  return (
    <Tag
      {...transferProps(restProps)}
      className={styles.span}
      style={{
        ...generateResponsiveCustomProperties(columns, 'column-span'),
        ...generateResponsiveCustomProperties(rows, 'row-span'),
      }}
    >
      {children}
    </Tag>
  );
};

export const GridSpanWithGlobalProps = withGlobalProps(GridSpan, 'GridSpan');

export default GridSpanWithGlobalProps;
