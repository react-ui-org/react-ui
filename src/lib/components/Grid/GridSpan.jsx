import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { generateResponsiveCustomProperties } from './_helpers/generateResponsiveCustomProperties';
import styles from './Grid.scss';

export const GridSpan = ({
  children,
  columns,
  id,
  rows,
  tag: Tag,
  ...other
}) => {
  if (!children) {
    return null;
  }

  return (
    <Tag
      id={id}
      className={styles.span}
      style={{
        ...generateResponsiveCustomProperties(columns, 'column-span'),
        ...generateResponsiveCustomProperties(rows, 'row-span'),
      }}
      {...other}
    >
      {children}
    </Tag>
  );
};

/* Breakpoints are easier to work with when ordered according to their value, not name. */
/* eslint-disable sort-keys */

GridSpan.defaultProps = {
  children: null,
  columns: undefined,
  id: undefined,
  rows: undefined,
  tag: 'div',
};

GridSpan.propTypes = {
  /**
   * Items to be aligned in the grid.
   */
  children: PropTypes.node,
  /**
   * Number of columns to span.
   */
  columns: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Number of rows to span.
   */
  rows: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
  /**
   * HTML tag to render. Can be any valid HTML tag of your choice, usually a
   * [block-level element](https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements).
   */
  tag: PropTypes.string,
};

export const GridSpanWithContext = withProviderContext(GridSpan, 'GridSpan');

export default GridSpanWithContext;
