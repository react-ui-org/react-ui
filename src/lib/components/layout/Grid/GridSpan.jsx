import PropTypes from 'prop-types';
import React from 'react';
import { generateResponsiveCustomProperties } from './helpers/generateResponsiveCustomProperties';
import styles from './Grid.scss';

export const GridSpan = ({
  children,
  columns,
  id,
  rows,
  ...other
}) => {
  if (!children) {
    return null;
  }

  return (
    <div
      id={id}
      className={styles.span}
      style={{
        ...generateResponsiveCustomProperties(columns, 'column-span'),
        ...generateResponsiveCustomProperties(rows, 'row-span'),
      }}
      {...other}
    >
      {children}
    </div>
  );
};

/* Breakpoints are easier to work with when ordered according to their value, not name. */
/* eslint-disable sort-keys */

GridSpan.defaultProps = {
  children: null,
  columns: undefined,
  id: undefined,
  rows: undefined,
};

GridSpan.propTypes = {
  /**
   * Items to be aligned in the grid.
   */
  children: PropTypes.node,
  /**
   * Number of columns to span.
   */
  columns: PropTypes.oneOf([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Number of rows to span.
   */
  rows: PropTypes.oneOf([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
      xl: PropTypes.number,
      xxl: PropTypes.number,
      xxxl: PropTypes.number,
    }),
  ]),
};

export default GridSpan;
