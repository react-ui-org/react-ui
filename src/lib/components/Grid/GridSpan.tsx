import React from 'react';
import { withGlobalProps } from '../../provider';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { transferProps } from '../_helpers/transferProps';
import { generateResponsiveCustomProperties } from './_helpers/generateResponsiveCustomProperties';
import styles from './Grid.scss';
import { GridSpanProps } from './Grid.types';

export const GridSpan: React.FunctionComponent<GridSpanProps> = ({
  children,
  columns = 1,
  rows = 1,
  tag: Tag = 'div',
  ...restProps
}) => {
  if (isChildrenEmpty(children)) {
    return null;
  }

  return (
    <Tag
      {...transferProps(restProps)}
      className={styles.span}
      style={{
        ...generateResponsiveCustomProperties(columns, 'column-span'),
        ...generateResponsiveCustomProperties(rows, 'row-span'),
      }}
    >
      {children}
    </Tag>
  );
};

export const GridSpanWithGlobalProps = withGlobalProps(GridSpan, 'GridSpan');

export default GridSpanWithGlobalProps;
