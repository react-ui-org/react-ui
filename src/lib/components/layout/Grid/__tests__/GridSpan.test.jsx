import React from 'react';
import { shallow } from 'enzyme';
import { GridSpan } from '../GridSpan';

describe('rendering', () => {
  it('renders correctly with no children', () => {
    const tree = shallow((
      <GridSpan />
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a single child', () => {
    const tree = shallow((
      <GridSpan>
        <div>content</div>
      </GridSpan>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <GridSpan>
        <div>content 1</div>
        <div>content 2</div>
        <div>content 3</div>
      </GridSpan>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with simple props', () => {
    const tree = shallow((
      <GridSpan
        columns={2}
        id="my-grid-span"
        rows={3}
      >
        <div>content</div>
      </GridSpan>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with complex props', () => {
    const tree = shallow((
      <GridSpan
        columns={{
          md: 1,
          xs: 2,
        }}
        id="my-grid-span"
        rows={{
          md: 4,
          xs: 3,
        }}
      >
        <div>content</div>
      </GridSpan>
    ));

    expect(tree).toMatchSnapshot();
  });
});
