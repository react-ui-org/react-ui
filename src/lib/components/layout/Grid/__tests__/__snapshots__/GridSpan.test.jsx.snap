// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="span"
  style={Object {}}
>
  <div>
    content
  </div>
</div>
`;

exports[`rendering renders correctly with complex props 1`] = `
<div
  className="span"
  id="my-grid-span"
  style={
    Object {
      "--rui-local-column-span-md": 1,
      "--rui-local-column-span-xs": 2,
      "--rui-local-row-span-md": 4,
      "--rui-local-row-span-xs": 3,
    }
  }
>
  <div>
    content
  </div>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="span"
  style={Object {}}
>
  <div>
    content 1
  </div>
  <div>
    content 2
  </div>
  <div>
    content 3
  </div>
</div>
`;

exports[`rendering renders correctly with no children 1`] = `""`;

exports[`rendering renders correctly with simple props 1`] = `
<div
  className="span"
  id="my-grid-span"
  style={
    Object {
      "--rui-local-column-span-xs": 2,
      "--rui-local-row-span-xs": 3,
    }
  }
>
  <div>
    content
  </div>
</div>
`;
